scapy/contrib/nativecansocket.py

```diff
@@ -0,0 +1,106 @@
+    desc = "read/write packets at a given CAN interface using PF_CAN sockets"
+
+    def __init__(self, iface=None, receive_own_messages=False, can_filters=None, remove_padding=True):
+        self.remove_padding = remove_padding
+        self.iface = conf.contribs['NativeCANSocket']['iface'] if iface is None else iface
+        self.ins = socket.socket(socket.PF_CAN, socket.SOCK_RAW, socket.CAN_RAW)
+        try:
+            self.ins.setsockopt(socket.SOL_CAN_RAW, socket.CAN_RAW_RECV_OWN_MSGS,
+                                struct.pack("i", receive_own_messages))
+        except Exception as exception:
+            Scapy_Exception("Could not modify receive own messages (%s)", exception)
+
+        if can_filters is None:
+            can_filters = [{
+                "can_id": 0,
+                "can_mask": 0
+            }]
+
+        can_filter_fmt = "={}I".format(2 * len(can_filters))
+        filter_data = []
+        for can_filter in can_filters:
+            filter_data.append(can_filter["can_id"])
+            filter_data.append(can_filter["can_mask"])
+
+        self.ins.setsockopt(socket.SOL_CAN_RAW,
+                            socket.CAN_RAW_FILTER,
+                            struct.pack(can_filter_fmt, *filter_data))
+
+        self.ins.bind((iface,))
+        self.outs = self.ins
+
+    def recv(self, x=CAN_FRAME_SIZE):
+        try:
+            pkt, sa_ll = self.ins.recvfrom(x)
+        except BlockingIOError:
+            warning("Captured no data, socket in non-blocking mode.")
+            return None
+        except socket.timeout:
+            warning("Captured no data, socket read timed out.")
+            return None
+        except OSError:
+            # something bad happened (e.g. the interface went down)
+            warning("Captured no data.")
+            return None
+
+        # need to change the byteoder of the first four bytes, required by the
+        # underlaying linux CAN frame format
+        pkt = struct.pack("<I12s", *struct.unpack(">I12s", pkt))
+        len = pkt[4]
+        canpkt = CAN(pkt[:len+8])
+        canpkt.time = get_last_packet_timestamp(self.ins)
+        if self.remove_padding:
+            return canpkt
+        else:
+            return canpkt / Padding(pkt[len + 8:])
+
+    def send(self, x):
+        try:
+            if hasattr(x, "sent_time"):
+                x.sent_time = time.time()
+
+            # need to change the byteoder of the first four bytes, required by the
+            # underlaying linux CAN frame format
+            bs = bytes(x)
+            bs = bs + b'\x00' * (CAN_FRAME_SIZE - len(bs))
+            bs = struct.pack("<I12s", *struct.unpack(">I12s", bs))
+            return SuperSocket.send(self, bs)
+        except socket.error as msg:
+            raise msg
+
+
+@conf.commands.register
+def srcan(pkt, iface=None, receive_own_messages=False, canfilter=None, *args, **kargs):
+    if "timeout" not in kargs:
+        kargs["timeout"] = -1
```

Comment:
Nice PR!
All good to me except this check, which is useless (see sendrecv.py)

> def sndrcv(pks, pkt, timeout=None ........
```
if timeout is not None and timeout < 0:
            timeout = None
```

I know it's everywhere in sendrecv.py, and I already deleted those in https://github.com/secdev/scapy/pull/1142

If you also rebase the PR, you'll have latest hotfix https://github.com/secdev/scapy/pull/1320 which fixes the `L2ListenTcpdump` failing bug (and will make your tests pass)

Otherwise PR seems ok to me !

Reply:
Done... I did a build on the server today, but unfortunatley, the hotfix didn't worked for me. See build 9595574 for reference.